firedrake/plot.py

```diff
@@ -0,0 +1,345 @@
+        points_1d = np.linspace(0, 1.0, num=num_points,
+                                dtype=float).reshape(-1, 1)
+        points = np.array(np.meshgrid(points_1d, points_1d)).T.reshape(-1, 2)
+    elif mesh.ufl_cell() == Cell('triangle'):
+        points_1d = np.linspace(0, 1.0, num=num_points,
+                                dtype=float).reshape(-1, 1)
+        points_1d_rev = np.fliplr([points_1d]).reshape(-1)
+        iu = np.triu_indices(num_points)
+        points = np.array(np.meshgrid(points_1d, points_1d_rev)).T[iu]
+    else:
+        raise RuntimeError("Unsupported functionality")
+    y_vals = _calculate_values(function, points, dimension)
+    x_vals = _calculate_values(mesh.coordinates, points, dimension)
+    return x_vals, y_vals
+
+
+def calculate_one_dim_points(function, num_points):
+    """Calculate a set of points for plotting for a one-dimension function as a
+    numpy array
+
+    :arg function: 1D function for plotting
+    :arg num_points: Number of points per element
+    """
+    x_vals, y_vals = _calculate_points(function, num_points, 1)
+    x_vals = x_vals.reshape(-1)
+    y_vals = y_vals.reshape(-1)
+    order = np.argsort(x_vals)
+    x_vals = x_vals[order]
+    y_vals = y_vals[order]
+    return np.array([x_vals, y_vals])
+
+
+def two_dimension_triangle_Z(function, num_sample_points):
+    """Calculate the triangulation and function values for a given 2D function
+
+    :arg function: 2D function
+    :arg num_sample_points: Number of sampling points
+    """
+
+    try:
+        from matplotlib.tri import Triangulation, UniformTriRefiner
+    except ImportError:
+        raise RuntimeError("Matplotlib not importable, is it installed?")
+    if function.function_space().mesh().cell() == Cell('triangle'):
+        x = np.array([0, 0, 1])
+        y = np.array([0, 1, 0])
+    elif function.function_space().mesh().cell() == Cell('quadrilateral'):
+        x = np.array([0, 0, 1, 1])
+        y = np.array([0, 1, 0, 1])
+    else:
+        raise RuntimeError("Unsupported Functionality")
+    base_tri = Triangulation(x, y)
+    refiner = UniformTriRefiner(base_tri)
+    tri = refiner.refine_triangulation(False, num_sample_points)
+    triangles = tri.get_masked_triangles()
+    x_ref = tri.x
+    y_ref = tri.y
+    num_verts = triangles.max() + 1
+    num_cells = function.function_space().cell_node_list.shape[0]
+    ref_points = np.dstack([x_ref, y_ref]).reshape(-1, 2)
+    z_vals = _calculate_values(function, ref_points, 2)
+    coords_vals = _calculate_values(function.function_space().
+                                    mesh().coordinates,
+                                    ref_points, 2)
+    Z = z_vals.reshape(-1)
+    X = coords_vals.reshape(-1, 2).T[0]
+    Y = coords_vals.reshape(-1, 2).T[1]
+    all_triangles = triangles.copy()
+    temp = triangles.copy()
+    for i in range(num_cells - 1):
+        temp = temp + num_verts
+        all_triangles = np.append(all_triangles, temp)
```

Comment:
There must be a cleverer way to do this that doesn't involve concatenating tiny arrays lots.  But I don't see it immediately.